libvips/foreign/foreign.c

```diff
@@ -1669,34 +1705,73 @@ vips__foreign_convert_saveable(VipsImage *in, VipsImage **ready,
 		in = out;
 	}
 
+	*ready = in;
+
+	return 0;
+}
+
+static void *
+vips_foreign_save_remove_metadata(VipsImage *image,
+	const char *field, GValue *value, void *user_data)
+{
+	VipsForeignKeep keep = *((VipsForeignKeep *) user_data);
+
+	// we are only interested in metadata
+	if (!vips_isprefix(field, "png-comment-") &&
+		!vips_isprefix(field, "magickprofile-") &&
+		strcmp(field, VIPS_META_IMAGEDESCRIPTION) != 0 &&
+		!vips_ispostfix(field, "-data"))
+		return NULL;
+
+	if ((strcmp(field, VIPS_META_EXIF_NAME) == 0 &&
+			(keep & VIPS_FOREIGN_KEEP_EXIF)) ||
+		(strcmp(field, VIPS_META_XMP_NAME) == 0 &&
+			(keep & VIPS_FOREIGN_KEEP_XMP)) ||
+		(strcmp(field, VIPS_META_IPTC_NAME) == 0 &&
+			(keep & VIPS_FOREIGN_KEEP_IPTC)) ||
+		(strcmp(field, VIPS_META_ICC_NAME) == 0 &&
+			(keep & VIPS_FOREIGN_KEEP_ICC)) ||
+		(keep & VIPS_FOREIGN_KEEP_OTHER))
+		return NULL;
+
+	if (!vips_image_remove(image, field))
+		return image;
+
+	return NULL;
+}
+
+int
+vips__foreign_update_metadata(VipsImage *in,
+	VipsForeignKeep keep)
+{
+	/* Rebuild exif from tags, if we'll be saving it.
+	 */
+	if ((keep & VIPS_FOREIGN_KEEP_EXIF) &&
+		vips__exif_update(in))
+		return -1;
```

Comment:
It looks like this wasn't called by every saver previously. See for example:

Git master:
```console
$ curl -LO https://github.com/kleisauke/net-vips/raw/master/tests/NetVips.Tests/images/йцук.jpg
$ vips copy йцук.jpg x.v
$ diff -u <(vipsheader йцук.jpg -a) <(vipsheader x.v -a)
```
<details>
  <summary>Diff</summary>

```diff
@@ -1,4 +1,4 @@
-йцук.jpg: 1024x768 uchar, 3 bands, srgb, jpegload
+x.v: 1024x768 uchar, 3 bands, srgb, jpegload
 width: 1024
 height: 768
 bands: 3
@@ -9,7 +9,7 @@
 yoffset: 0
 xres: 7.08661
 yres: 7.08661
-filename: йцук.jpg
+filename: x.v
 vips-loader: jpegload
 jpeg-multiscan: 0
 jpeg-chroma-subsample: 4:2:0
@@ -64,3 +64,4 @@
 jpeg-thumbnail-data: 3160 bytes of binary data
 orientation: 1
 icc-profile-data: 1352 bytes of binary data
+vips-sequential: 1
```
</details>



This PR:
```console
$ curl -LO https://github.com/kleisauke/net-vips/raw/master/tests/NetVips.Tests/images/йцук.jpg
$ vips copy йцук.jpg x.v
$ diff -u <(vipsheader йцук.jpg -a) <(vipsheader x.v -a)
```

<details>
  <summary>Diff</summary>

```diff
@@ -1,4 +1,4 @@
-йцук.jpg: 1024x768 uchar, 3 bands, srgb, jpegload
+x.v: 1024x768 uchar, 3 bands, srgb, jpegload
 width: 1024
 height: 768
 bands: 3
@@ -9,11 +9,13 @@
 yoffset: 0
 xres: 7.08661
 yres: 7.08661
-filename: йцук.jpg
+filename: x.v
 vips-loader: jpegload
 jpeg-multiscan: 0
 jpeg-chroma-subsample: 4:2:0
-exif-data: 6626 bytes of binary data
+icc-profile-data: 1352 bytes of binary data
+vips-sequential: 1
+exif-data: 4302 bytes of binary data
 resolution-unit: in
 exif-ifd0-ImageDescription:                                 (                               , ASCII, 32 components, 32 bytes)
 exif-ifd0-Make: Canon (Canon, ASCII, 6 components, 6 bytes)
@@ -45,12 +47,12 @@
 exif-ifd2-MeteringMode: 5 (Pattern, Short, 1 components, 2 bytes)
 exif-ifd2-Flash: 16 (Flash did not fire, compulsory flash mode, Short, 1 components, 2 bytes)
 exif-ifd2-FocalLength: 22500/1000 (22.5 mm, Rational, 1 components, 8 bytes)
-exif-ifd2-MakerNote: 2290 bytes undefined data (2290 bytes undefined data, Undefined, 2290 components, 2290 bytes)
+exif-ifd2-MakerNote: 6 bytes undefined data (6 bytes undefined data, Undefined, 6 components, 6 bytes)
 exif-ifd2-UserComment:  (, Undefined, 264 components, 264 bytes)
 exif-ifd2-FlashpixVersion: FlashPix Version 1.0 (FlashPix Version 1.0, Undefined, 4 components, 4 bytes)
 exif-ifd2-ColorSpace: 1 (sRGB, Short, 1 components, 2 bytes)
-exif-ifd2-PixelXDimension: 3648 (3648, Short, 1 components, 2 bytes)
-exif-ifd2-PixelYDimension: 2736 (2736, Short, 1 components, 2 bytes)
+exif-ifd2-PixelXDimension: 1024 (1024, Short, 1 components, 2 bytes)
+exif-ifd2-PixelYDimension: 768 (768, Short, 1 components, 2 bytes)
 exif-ifd2-FocalPlaneXResolution: 3648000/292 (12493.151, Rational, 1 components, 8 bytes)
 exif-ifd2-FocalPlaneYResolution: 2736000/219 (12493.151, Rational, 1 components, 8 bytes)
 exif-ifd2-FocalPlaneResolutionUnit: 2 (Inch, Short, 1 components, 2 bytes)
@@ -63,4 +65,3 @@
 exif-ifd2-SceneCaptureType: 0 (Standard, Short, 1 components, 2 bytes)
 jpeg-thumbnail-data: 3160 bytes of binary data
 orientation: 1
-icc-profile-data: 1352 bytes of binary data
```

</details>

So, when saving the image in vips format, this PR corrects the `exif-ifd2-PixelXDimension` / `exif-ifd2-PixelYDimension` tags and changes the `exif-ifd2-MakerNote` tag (probably because we don't use the `EXIF_DATA_OPTION_DONT_CHANGE_MAKER_NOTE` option).

I'm not sure if this new behaviour is desired. It caused a test failure in NetVips (which I was able to fix by changing the test fixture, see e.g. commit https://github.com/kleisauke/net-vips/commit/1c4acefc8288ec318bd58ce2a144fa731785ce9f).

Reply:
Huh that's true. I think this only happens for save via the `foreign/` interface, if you use the vips writer in `iofuncs/` directly, this probably won't happen, which is even more confusing.

The alternative is to move the exif update out of the base saver and only call it from format savers that have direct EXIF support (what we had before). On the other hand, I think this only happens with vips format, which is maybe not very important.

What do you think?

Reply:
It probably only happens with the vips format, so indeed, let's leave it as is.

Could invoking `vips__exif_update()` potentially impact performance? For instance, formats that don't support Exif at all (e.g. `gifsave`) would not need this.

Reply:
Yes, there's a performance hit. My instinct says it's negligible, but I've not done any experiments.

Reply:
I also think it's negligible. If anyone complains, we could set the "keep" flag automatically based on which metadata is supported by the saver.